src/diffpy/utils/parsers/loadmetafile.py

```diff
@@ -0,0 +1,212 @@
+
+import pathlib
+import json
+
+# FIXME: add support for yaml, xml
+supported_formats = ['.json']
+
+
+def load_PDF_into_db(dbname, pdfname, hddata: dict, rv: list, show_path=True):
+    """Load an entry consisting of PDF header and base data into a database file.
+
+    Requires hdata and rv generated from loadData.
+
+    dbname          -- name of the database file to load an entry into.
+    pdfname         -- name of the PDF file.
+    hddata          -- Dictionary of PDF metadata generated by loadData.
+    rv              -- List of PDF (r, gr) pairs generated by loadData.
+    show_path       -- include a PDF_path element in the database entry (default True).
+
+    Returns the dictionary loaded from/into the updated database file.
+    """
+    # new file or update
+    existing = False
+    if pathlib.Path.is_file(dbname):
+        existing = True
+
+    # collect entry
+    with open(pdfname, 'r') as grfile:
+        data = {}
+
+        # add path
+        grpath = grfile.name
+        if show_path:
+            data.update({'PDF_path': grpath})
+
+        # add r, gr, and header metadata
+        data.update({'r': list(rv[:, 0]), 'gr': list(rv[:, 1])})
+        data.update(hddata)
+
+        # parse name using pathlib and generate json entry
+        name = pathlib.Path(grpath).name
+        entry = {name: data}
+
+    # check if supported type
+    extension = pathlib.Path(dbname).suffix
+    if extension not in supported_formats:
+        raise Exception(f"Format of {dbname} is not supported.")
+
+    # json
+    if extension == '.json':
+        # dump if non-existing
+        if not existing:
+            with open(dbname, 'w') as jsonfile:
+                pdfs = entry  # for return
+                json.dump(pdfs, jsonfile, indent=2)
+
+        # update if existing
+        else:
+            with open(dbname, 'r') as json_read:
+                pdfs = json.load(json_read)
+                pdfs.update(entry)
+            with open(dbname, 'w') as json_write:
+                # dump to string first for formatting
+                json.dump(pdfs, json_write, indent=2)
+
+    return pdfs
+
+
+def markup_PDF(hddata: dict, rv: list, muname=None):
+    # FIXME: may be better suited for REST API package, not diffpy.utils
+    """Put PDF file information into a dictionary.
+
+    hddata  -- Dictionary of metadata.
+    rv      -- List of (r, gr) pairs.
+    muname  -- file to save into (default None, no saving occurs).
+
+    Returns the dictionary loaded from/into markup file.
+    """
+
+    # gather data
+    data = {}
+    data.update({'r': list(rv[:, 0]), 'gr': list(rv[:, 1])})
+    data.update(hddata)
+
+    # return directly
+    if muname is None:
+        return data
+
+    # save to disk when enabled
+    extension = pathlib.Path(muname).suffix
+    if extension not in supported_formats:
+        raise Exception(f"Format of {muname} is not supported.")
```

Comment:
we could also just let loadData decide what it can't read and it can raise an exception? Otherwise this can work?

Reply:
Exception raised when trying to put our dictionary in a markup language that we do not have write support for. Currently, I have only put json support for the REST API, but we can later extend to yaml or xml.

Separate from loadData which would deal with read rather than write support.

Reply:
sorry, I misunderstood.   A clearer docstring maybe would have helped there.

When we move things from memory to more permanent storage we call it serialization.   It is better to use this language.  I am not sure if json is strictly a markup language, it is more like a set of standards for serializing code, so I might stay away from that language too.   The function seems to be doing two things.  One is taking data and putting into a dictionary format and passing it back.  The other is that it is serializing it to disc.  I recommend to split these two functionalities into two separate functions.  One creates the data object (in this case a dict) and the other does serialization, in this case to json.   It is clearer if they are in different functions.

Reply:
Thanks for informing that it is called serialization. Changed the nomenclature. Was a bit concerned about what term to use. Very helpful o.O :)